### PR TITLE
Us022 disable respondent enrolment

### PR DESCRIPTION
**Pull request by andrewmil**
Added an endpoint to change a party's enrolment status for a given survey. When we do this we also post a case event for each relevant case

Works with other PR's:
[response-operations-ui](https://github.com/ONSdigital/response-operations-ui/pull/123)
[ras-backstage](https://github.com/ONSdigital/ras-backstage/pull/96)
[ras-frontstage-api](https://github.com/ONSdigital/ras-frontstage-api/pull/60)
[rm-case](https://github.com/ONSdigital/rm-case-service/pull/28)
[rm-casesvc-api](https://github.com/ONSdigital/rm-casesvc-api/pull/13)

[Confluence](https://digitaleq.atlassian.net/wiki/spaces/RASB/pages/248545529/US022+Disable+respondent+enrolment+to+a+survey)
[Trello](https://trello.com/c/1ez85L0y/51-us022-int-disable-respondents-enrolment-to-a-survey-l)

### CODE REVIEW

File: test/test_respondent_controller.py [hunk 1]
```diff
@@ -680,6 +698,67 @@ def test_post_add_survey_no_business_raise_ras_error(self):
...
+        token = self.generate_valid_token_from_email(db_respondent.email_address)
+        self.put_email_verification(token, 200)
+        request_json = {
+            'respondent_party_id': self.mock_respondent_with_id['id'],
+            'business_id': '3b136c4b-7a14-4904-9e01-13364dd7b972',
+            'survey_id': '02b9c366-7397-42f7-942a-76dc5876d86d',
+            'change_flag': 'woafouewbhouGFHEPIW0'
+        }
+        self.put_enrolment_status(request_json, 500)
+
```
Comment (gemmai95) [inline test/test_respondent_controller.py:776]: I think it needs a test for if the respondent we are trying to change the enrolment status of doesn't exist.

**Review by gemmai95 — CHANGES_REQUESTED** [review]
Do we know what the requirements are for being able to change an enrolment in pending state to enabled or disabled? Also can we change an enrolment back to pending as I think the current implementation I think allows this?

**Comment by gemmai95** [comment]
[US022: INT Disable respondents enrolment to a survey [L]](https://trello.com/c/1ez85L0y/51-us022-int-disable-respondents-enrolment-to-a-survey-l)

**Comment by gemmai95** [comment]
[US020: INT Suspend respondents account (L) - Needed for R17](https://trello.com/c/sDXLtTMy/56-us020-int-suspend-respondents-account-l-needed-for-r17)

File: ras_party/controllers/account_controller.py [hunk 2]
```diff
@@ -526,3 +556,45 @@ def post_case_event(case_id, party_id, category='Default category message', desc
     logger.info('Case service responded with', status=response.status_code)
     response.raise_for_status()
     return response.json()
+
+
+def request_cases_for_respondent(respondent_id):
+    case_url = f'{current_app.config["RAS_CASE_SERVICE"]}/cases/partyid/{respondent_id}'
+    logger.info('GET', url=case_url)
```
Comment (jwgwalton) [inline ras_party/controllers/account_controller.py]: The logs are going to get pretty cluttered with INFO's logging the various calls.

File: ras_party/controllers/account_controller.py [hunk 3]
```diff
@@ -526,3 +556,45 @@ def post_case_event(case_id, party_id, category='Default category message', desc
...
+def request_collection_exercises_for_survey(survey_id):
+    case_url = f'{current_app.config["RAS_COLLEX_SERVICE"]}/collectionexercises/survey/{survey_id}'
+    logger.info('GET', url=case_url)
+    response = Requests.get(case_url)
+    logger.info('Survey service responded with', status=response.status_code)
+    response.raise_for_status()
+    return response.json()
+
+
+def get_matching_cases(survey_id, business_id, respondent_id):
```
Comment (jwgwalton) [inline ras_party/controllers/account_controller.py]: get_cases_for_collex

File: test/test_respondent_controller.py (hunk 1 above)
Comment (jwgwalton) [reply test/test_respondent_controller.py:776]: +1

**Review by jwgwalton — DISMISSED** [review]

File: ras_party/controllers/account_controller.py (hunk 2 above)
Comment (andrewmil) [reply ras_party/controllers/account_controller.py]: I'll change some of this to debug and improve the logging a bit, these messages aren't that useful as they are now

File: test/test_respondent_controller.py (hunk 1 above)
Comment (andrewmil) [reply test/test_respondent_controller.py:776]: Will add this in :+1:

File: ras_party/controllers/account_controller.py [hunk 4]
```diff
@@ -132,6 +133,38 @@ def post_respondent(party, tran, session):
...
+
+    enrolment = query_enrolment_by_survey_business_respondent(respondent_id=respondent.id,
+                                                              business_id=business_id,
+                                                              survey_id=survey_id,
+                                                              session=session)
+    enrolment.status = change_flag
+
+    category = 'DISABLE_RESPONDENT_ENROLMENT' if change_flag == 'DISABLED' else 'ENABLE_RESPONDENT_ENROLMENT'
+    for case in get_cases_for_collection_exercise(survey_id, business_id, respondent_id):
+        post_case_event(case['id'], business_id, category=category)
```
Comment (tej99) [inline ras_party/controllers/account_controller.py]: The desc param is left as the default, do you want the event description to be 'Default description message'?

Comment (tej99) [inline ras_party/controllers/account_controller.py]: The param 'desc' is left as the default, do you want the event description to be 'Default description message'?

**Review by tej99 — APPROVED** [review]
LGTM just one small thing

File: ras_party/controllers/account_controller.py [hunk 5]
```diff
@@ -521,8 +556,49 @@ def post_case_event(case_id, party_id, category='Default category message', desc
...
+    logger.debug('Retrieving collection exercises for survey', survey_id=survey_id)
+    url = f'{current_app.config["RAS_COLLEX_SERVICE"]}/collectionexercises/survey/{survey_id}'
+    response = Requests.get(url)
+    response.raise_for_status()
+    logger.debug('Successfully retrieved collection exercises for survey', survey_id=survey_id)
+    return response.json()
+
+
+def get_cases_for_collection_exercise(survey_id, business_id, respondent_id):
+    collection_exercises = request_collection_exercises_for_survey(survey_id)
```
Comment (gemmai95) [inline ras_party/controllers/account_controller.py]: Does this function need some logging?

**Review by gemmai95 — APPROVED** [review]

**Pull request merged**